File: windows/src/GoalMaker.Core/Planning/PeriodFacts.cs
namespace GoalMaker.Core.Planning;

/// <summary>
/// What a review period looked like, as the reactive prompts read it (docs/reviews.md): the tasks done
/// against the <see cref="AverageDone"/> of the periods before it, the goals with where they stand
/// against where they should be, the habits with the periods they missed, and the tasks that kept moving.
/// </summary>
public sealed record PeriodFacts
{
    public int DoneTasks { get; init; }

    public double AverageDone { get; init; }

    public IReadOnlyList<GoalFact> Goals { get; init; } = [];

    public IReadOnlyList<HabitFact> Habits { get; init; } = [];

    public IReadOnlyList<TaskFact> Tasks { get; init; } = [];

    /// <summary>A goal's progress (<paramref name="Fraction"/>, 0 to 1) against the expected share of its period gone by.</summary>
    public sealed record GoalFact(string Title, double Fraction, double Expected);

    /// <summary>A habit's missed periods out of the ones it had, and the streak it is on now.</summary>
    public sealed record HabitFact(string Name, int Missed, int Periods, int Streak);

    /// <summary>A task and how many times it was moved to another day.</summary>
    public sealed record TaskFact(string Title, int Moves);
}

File: android/app/src/main/kotlin/com/goalmaker/app/domain/planning/PeriodFacts.kt
package com.goalmaker.app.domain.planning

/**
 * What a review period looked like, as the reactive prompts read it (docs/reviews.md): the tasks
 * done against the [averageDone] of the periods before it, the goals with where they stand against
 * where they should be, the habits with the periods they missed, and the tasks that kept moving.
 */
data class PeriodFacts(
    val doneTasks: Int = 0,
    val averageDone: Double = 0.0,
    val goals: List<GoalFact> = emptyList(),
    val habits: List<HabitFact> = emptyList(),
    val tasks: List<TaskFact> = emptyList(),
) {
    /** A goal's progress ([fraction], 0 to 1) against the [expected] share of its period gone by. */
    data class GoalFact(val title: String, val fraction: Double, val expected: Double)

    /** A habit's [missed] periods out of [periods], and the streak it is on now. */
    data class HabitFact(val name: String, val missed: Int, val periods: Int, val streak: Int)

    /** A task and how many times it was moved to another day. */
    data class TaskFact(val title: String, val moves: Int)
}
